### PR TITLE
[WIP] begin building a config flags struct

### PR DESCRIPTION
**Pull request by juanvallejo**
Work in progress...

Opening this now to gather feedback on whether this is the design we want.

cc @soltysh @deads2k

### CODE REVIEW

File: pkg/kubectl/cmd/util/config_flags.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+	overrides := &clientcmd.ConfigOverrides{ClusterDefaults: clientcmd.ClusterDefaults}
+	clientConfig := clientcmd.NewInteractiveDeferredLoadingClientConfig(loadingRules, overrides, os.Stdin)
+
+	return clientConfig.ClientConfig()
+}
+
+// ToDiscoveryClient returns a CachedDiscoveryInterface using a computed RESTConfig.
+// Expects the AddFlags method to have been called.
+func (f *ConfigFlags) ToDiscoveryClient() (discovery.CachedDiscoveryInterface, error) {
```
Comment (juanvallejo) [inline pkg/kubectl/cmd/util/config_flags.go]: @deads2k do we want this method? Do we want to just provide a discovery factory? It appears that we need one of those [to be able to obtain a client cache](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/clientcache.go#L32), to then [obtain a kube clientset](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/factory_client_access.go#L204).

Unless we want a different approach for obtaining a clientset from this struct?

**Comment by juanvallejo** [comment]
Not yet sure which other methods we want to tack on to this. Added the `ToRESTConfig` method for now.

File: pkg/kubectl/cmd/util/config_flags.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/homedir"
+
+	"k8s.io/kubernetes/pkg/kubectl/util/transport"
+)
+
+// ConfigFlags composes the set of values necessary
+// for obtaining a REST client config
+type ConfigFlags struct {
+	CacheDir   *string
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/config_flags.go]: we expose this for some reason?

Comment (deads2k) [inline pkg/kubectl/cmd/util/config_flags.go]: default this one to nil so we don't expose it by default.  I mean who really needs to change that?

File: pkg/kubectl/cmd/util/config_flags.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/homedir"
+
+	"k8s.io/kubernetes/pkg/kubectl/util/transport"
+)
+
+// ConfigFlags composes the set of values necessary
+// for obtaining a REST client config
+type ConfigFlags struct {
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/config_flags.go]: you need a lot more flags.  token, ca, server, insecure, client-crt, username, etc

File: pkg/kubectl/cmd/util/config_flags.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	// use the standard defaults for this client command
+	// DEPRECATED: remove and replace with something more accurate
+	loadingRules.DefaultClientConfig = &clientcmd.DefaultClientConfig
+
+	if f.KubeConfig != nil {
+		loadingRules.ExplicitPath = *f.KubeConfig
+	}
+
+	overrides := &clientcmd.ConfigOverrides{ClusterDefaults: clientcmd.ClusterDefaults}
+	clientConfig := clientcmd.NewInteractiveDeferredLoadingClientConfig(loadingRules, overrides, os.Stdin)
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/config_flags.go]: We want this part to be configurable based on whether we allow username/password or not.  If we don't, this should be non-interactive.

File: pkg/kubectl/cmd/util/config_flags.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+	return clientConfig.ClientConfig()
+}
+
+// ToDiscoveryClient returns a CachedDiscoveryInterface using a computed RESTConfig.
+// Expects the AddFlags method to have been called.
+func (f *ConfigFlags) ToDiscoveryClient() (discovery.CachedDiscoveryInterface, error) {
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/config_flags.go]: This method is fine.

File: pkg/kubectl/cmd/util/config_flags.go [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+	BearerToken      *string
+	Impersonate      *string
+	ImpersonateGroup *string
+	Username         *string
+	Password         *string
+	Timeout          *string
+
+	// DisableUsernamePassword allows for the username and password config
+	// options to be disabled, in turn using a non-interactive client config loader.
+	DisableUsernamePassword bool
```
Comment (soltysh) [inline pkg/kubectl/cmd/util/config_flags.go]: I'd turn this into a struct's private member you can only set once when initiating the struct with `NewConfigFlags`

**Comment by soltysh** [comment]
Start wiring that into one of the commands and you'll see how's this working for you.

File: pkg/kubectl/cmd/util/config_flags.go [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/homedir"
+
+	"k8s.io/kubernetes/pkg/kubectl/util/transport"
+)
+
+const (
+	FlagClusterName      = "cluster"
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/config_flags.go]: don't expose these

File: pkg/kubectl/cmd/util/config_flags.go (hunk 6 above)
Comment (deads2k) [reply pkg/kubectl/cmd/util/config_flags.go]: > I'd turn this into a struct's private member you can only set once when initiating the struct with NewConfigFlags

Doing that means you can't construct a custom version of this.  YOu'd have to call the constructor and get a bunch of values already set.

File: pkg/kubectl/cmd/util/config_flags.go (hunk 1 above)
Comment (deads2k) [reply pkg/kubectl/cmd/util/config_flags.go]: > @deads2k do we want this method? Do we want to just provide a discovery factory? It appears that we need one of those to be able to obtain a client cache, to then obtain a kube clientset.

the cache is gone

File: pkg/kubectl/cmd/util/config_flags.go [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+	if f.Timeout != nil {
+		cmd.Flags().StringVar(f.Timeout, FlagTimeout, *f.Timeout, "The length of time to wait before giving up on a single server request. Non-zero values should contain a corresponding time unit (e.g. 1s, 2m, 3h). A value of zero means don't timeout requests.")
+	}
+
+	if f.EmbedCerts != nil {
+		// TODO: there is no flag-info present for this flag in clientcmd/overrides.go
+	}
+}
+
+func NewConfigFlags(disableUsernamePassword bool) *ConfigFlags {
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/config_flags.go]: Ah, ok.  Two flavors here. `NewWithDeprecatedPasswordFlag` that has non-nil, username and password.  and `NewConfigFlags()` which has nil username and password

File: pkg/kubectl/cmd/util/config_flags.go [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+	loadingRules.DefaultClientConfig = &clientcmd.DefaultClientConfig
+
+	if f.KubeConfig != nil {
+		loadingRules.ExplicitPath = *f.KubeConfig
+	}
+
+	overrides := &clientcmd.ConfigOverrides{ClusterDefaults: clientcmd.ClusterDefaults}
+
+	var clientConfig clientcmd.ClientConfig
+	if f.DisableUsernamePassword {
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/config_flags.go]: choose this based on a non-nil password string ptr

File: pkg/kubectl/cmd/util/config_flags.go [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+		// TODO: there is no flag-info present for this flag in clientcmd/overrides.go
+	}
+}
+
+func NewConfigFlags(disableUsernamePassword bool) *ConfigFlags {
+	kubeConfig := ""
+	insecureVal := "false"
+	timeout := "0"
+
+	return &ConfigFlags{
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/config_flags.go]: Most of the flags should be set to non-nil values.

File: pkg/kubectl/cmd/util/config_flags.go (hunk 6 above)
Comment (deads2k) [inline pkg/kubectl/cmd/util/config_flags.go]: this looks unnecessary anyway

File: pkg/kubectl/cmd/util/config_flags.go [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	// use the standard defaults for this client command
+	// DEPRECATED: remove and replace with something more accurate
+	loadingRules.DefaultClientConfig = &clientcmd.DefaultClientConfig
+
+	if f.KubeConfig != nil {
+		loadingRules.ExplicitPath = *f.KubeConfig
+	}
+
+	overrides := &clientcmd.ConfigOverrides{ClusterDefaults: clientcmd.ClusterDefaults}
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/config_flags.go]: seems like this is missing the flag values.

**Comment by juanvallejo** [comment]
@soltysh @deads2k thanks, review comments addressed. Will bind flags at command level

File: pkg/kubectl/cmd/util/config_flags.go [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+	flagClusterName      = "cluster"
+	flagAuthInfoName     = "user"
+	flagContext          = "context"
+	flagNamespace        = "namespace"
+	flagAPIServer        = "server"
+	flagInsecure         = "insecure-skip-tls-verify"
+	flagCertFile         = "client-certificate"
+	flagKeyFile          = "client-key"
+	flagCAFile           = "certificate-authority"
+	flagEmbedCerts       = "embed-certs"
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/config_flags.go]: remove

Comment (juanvallejo) [reply pkg/kubectl/cmd/util/config_flags.go]: Thanks, done

**Comment by juanvallejo** [comment]
@deads2k moving config_flags.go to `pkg/kubectl/genericclioptions` would cause an import cycle as the `ToDiscoveryClientMethod` would depend on the `pkg/kubectl/cmd/util` pakckage in order to instantiate the discovery client, and the `pkg/kubectl/cmd/util` package would depend on genericclioptions in order to receive a RESTClientGetter

File: pkg/kubectl/cmd/util/config_flags.go [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	// use the standard defaults for this client command
+	// DEPRECATED: remove and replace with something more accurate
+	loadingRules.DefaultClientConfig = &clientcmd.DefaultClientConfig
+
+	if f.KubeConfig != nil {
+		loadingRules.ExplicitPath = *f.KubeConfig
+	}
+
+	flagNames := clientcmd.RecommendedConfigOverrideFlags("")
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/config_flags.go]: why is this here?

File: pkg/kubectl/cmd/util/config_flags.go [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+	}
+
+	return clientConfig
+}
+
+// ToDiscoveryClient implements RESTClientGetter.
+// Expects the AddFlags method to have been called.
+// Returns a CachedDiscoveryInterface using a computed RESTConfig.
+func (f *ConfigFlags) ToDiscoveryClient() (discovery.CachedDiscoveryInterface, error) {
+	if f.CacheDir == nil {
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/config_flags.go]: don't do this here.  just let the flag binding win for now.

File: pkg/kubectl/cmd/util/config_flags.go [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+	}
+	if f.Namespace != nil {
+		flags.StringVarP(f.Namespace, flagNamespace, "n", *f.Namespace, "If present, the namespace scope for this CLI request")
+	}
+	if f.Context != nil {
+		flags.StringVar(f.Context, flagContext, *f.Context, "The name of the kubeconfig context to use")
+	}
+
+	if f.APIServer != nil {
+		flags.StringVar(f.APIServer, flagAPIServer, *f.APIServer, "The address and port of the Kubernetes API server")
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/config_flags.go]: don't we support `-s` too?

File: pkg/kubectl/cmd/util/config_flags.go [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+	}
+
+	return rest.RESTClientFor(clientConfig)
+}
+
+// ToRESTConfig implements RESTClientGetter.
+// Returns a REST client configuration based on a provided path
+// to a .kubeconfig file, loading rules, and config flag overrides.
+// Expects the AddFlags method to have been called.
+func (f *ConfigFlags) ToRESTConfig() (*rest.Config, clientcmd.ClientConfig, error) {
```
[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/kubectl/cmd/util/config_flags.go]: > I ended up returning clientcmd.ClientConfig as well since the factory_client_access.go#DefaultNamespace method uses it to lookup the namespace

That'll be a different method.

**Comment by juanvallejo** [comment]
Thanks for the review, will open PR against upstream
